### PR TITLE
AKCORE-3: Add ShareGroupHeartbeat RPC which is marked as an unstable RPC , supports authorization and group.share.enable config

### PR DESCRIPTION
**Pull request by adixitconfluent**
**About**
This PR implements ShareGroupHeartbeat RPC which is marked as an unstable RPC , supports authorization and group.share.enable config
Reference - [AKCORE-3](https://confluentinc.atlassian.net/browse/AKCORE-3)

The implementation to support ShareGroupHeartbeat RPC is not done yet in the Group Coordinator so I have mentioned it as a [TODO](https://github.com/confluentinc/kafka/pull/1003/files#diff-cc056b4960ededba37a438b1454f0f3c5ff5e8ad5e6d2ec9a08e813ca056ffebR3819) which needs to be followed once the implementation of Group Coordinator is complete. Right now we give an INVALID_CONFIG exception for that code scenario. 

**Build**
The build is passing as shown in the screenshot
Screenshot 2024-02-05 at 11 48 27 AM

The check styles is passing as shown in the screenshot
Screenshot 2024-02-05 at 11 50 05 AM

**Testing**
Added 3 unit tests to demonstrate the added code functionality -

1. `testShareGroupHeartbeatReturnsUnsupportedVersion` 
Screenshot 2024-02-05 at 1 10 56 PM

2. `testShareGroupHeartbeatRequest`  - The current test checks for INVALID_CONFIG exception, however once the RPC is supported in Group Coordinator, this [TODO](https://github.com/confluentinc/kafka/pull/1003/files#diff-d418e862d65bd68453ec359a3b4f08e38be8948b4d772909ece6ff2a07e5f714R7283) should be followed (I have tested the commented out test by using Andrew's prototype code for ShareGroupHeartbeat API support in Group Coordinator)
Screenshot 2024-02-05 at 1 11 09 PM

3. `testShareGroupHeartbeatRequestAuthorizationFailed` 
Screenshot 2024-02-05 at 1 11 20 PM

4. `testShareGroupHeartbeatIsAccessibleWhenEnabled` 
Screenshot 2024-02-06 at 6 30 36 PM

Note - I have run the tests for approx. 10 times to make sure they are not flaky.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

[AKCORE-3]: https://confluentinc.atlassian.net/browse/AKCORE-3?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: config/kraft/broker.properties
Comment (adixitconfluent) [reply config/kraft/broker.properties]: Sure, I have reverted the file change

File: config/kraft/server.properties
Comment (adixitconfluent) [reply config/kraft/server.properties]: Sure, I have reverted the file change

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:301]:
```suggestion
  /** Share Group Configurations **/
```

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:794]: Any reason of not defining the placeholder as above i.e. `Configurations **/`, if not then:
```suggestion
  /** Share Group Configurations * */
```

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1820]: Do we need to define the block `/** Share Group Configurations **/` here as well?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (apoorvmittal10) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:7310]: Instead of commenting the code, you can mark the test as @Disabled with a TODO: Enable once xyz changes are merged.

**Review by apoorvmittal10 — CHANGES_REQUESTED** [review]
Thanks for the PR @adixitconfluent. Left some comments, also can you please add tests for ....Request|Response.java classes as well.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm

File: core/src/main/scala/kafka/network/RequestConvertToJson.scala
Comment (adixitconfluent) [reply core/src/main/scala/kafka/network/RequestConvertToJson.scala:106]: understood, didn't know about the order, will make the corresponding change

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3812]: ack

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3835]: since we are already sending Errors.UNSUPPORTED_VERSION in the case when config.isShareGroupEnabled is false ([reference](https://github.com/confluentinc/kafka/pull/1003/files#diff-cc056b4960ededba37a438b1454f0f3c5ff5e8ad5e6d2ec9a08e813ca056ffebR3813)), I used a different error here and from the list of errors, this one seemed more suitable than others. Happy to change if you disagree

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1820]: I added this comment to make it synchronous with a [comment](https://github.com/confluentinc/kafka/pull/1003/files#diff-cbe6a8b71b05ed22cf09d97591225b588e9fca6caaf95d3b34a43262cfd23aa6R1804) when we enable the new group coordinator. Added the block `/** Share Group Configurations **/` there

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:794]: ack

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (adixitconfluent) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:7310]: thanks for pointing this out. I'll make the coorresponding change

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1165]: Agreed. I had initially put the configuration in both server.properties and broker.properties and thought it would be internal but Andrew pointed out that it shouldn't be the case. I'll make the change

File: clients/src/main/resources/common/message/ShareGroupHeartbeatResponse.json
Comment (adixitconfluent) [reply clients/src/main/resources/common/message/ShareGroupHeartbeatResponse.json:38]: Not sure if it followed everywhere. I see in [ConsumerGroupHeartbeatRequest.json](https://github.com/apache/kafka/blob/trunk/clients/src/main/resources/common/message/ConsumerGroupHeartbeatResponse.json#L41), it is mentioned as "id" in about. But it is followed in majority of the places as you pointed out, so I'll make the change

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3835]: Doesn't matter much though as the code will be changed shortly, but again it's better to use UNSUPPORTED_VERSION.

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1821]: The comment you reference actually talks about 2 cases but that's not the behaviour here. I would have removed it but I ll leave it to you to decide.

**Comment by adixitconfluent** [comment]
> Thanks for the PR @adixitconfluent. Left some comments, also can you please add tests for ....Request|Response.java classes as well.

Hi @apoorvmittal10, Regarding adding comments for ShareGroupHeartbeatRequest | ShareGroupHeartbeatResponse.java, these classes are tested through the 3 tests mentioned in `KafkaApisTest.scala`. I see ConsumerGroupHeartbeatRequestTest.java but most of those tests use group coordinator functionality except using UNSUPPORTED_VERSION, so I can add 1 for that. I will add a test similar to that for share groups.

File: core/src/main/scala/kafka/server/KafkaConfig.scala
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1821]: Sure, I'll go ahead and remove it

File: clients/src/main/resources/common/message/ShareGroupHeartbeatResponse.json
Comment (AndrewJSchofield) [reply clients/src/main/resources/common/message/ShareGroupHeartbeatResponse.json:58]: Probably a sensible idea. I've not done it before but there are some examples already in Kafka for this kind of thing.

**Review by apoorvmittal10 — APPROVED** [review]
lgtm

**Pull request merged**
